File: src/ts/radio/waitToken.ts
/**
 * 轻量级的请求-响应等待令牌。
 */

export class WaitToken<T = any> {
    waitCmd: string;
    timeBaseId: number;
    response: T | null = null;
    processedResponse: any = null;

    private _resolve?: (value: any) => void;
    private _reject?: (reason?: any) => void;
    private _promise: Promise<any>;
    private _done: boolean = false;
    private _postProcessor?: (data: T) => any;

    constructor(waitCmd: string, timeBaseId: number) {
        this.waitCmd = waitCmd;
        this.timeBaseId = timeBaseId;
        this._promise = new Promise((resolve, reject) => {
            this._resolve = resolve;
            this._reject = reject;
        });
    }

    get done(): boolean {
        return this._done;
    }

    setPostProcessor(processor: (data: T) => any): WaitToken<T> {
        this._postProcessor = processor;
        return this;
    }

    private _applyPostProcessor(data: T): any {
        if (this._postProcessor) {
            this.processedResponse = this._postProcessor(data);
        } else {
            this.processedResponse = data;
        }
        return this.processedResponse;
    }

    complete(data: T): void {
        this.response = data;
        const result = this._applyPostProcessor(data);
        this._done = true;
        if (this._resolve) {
            this._resolve(result);
        }
    }

    async wait(timeoutMs: number = 3000): Promise<any> {
        const timeoutPromise = new Promise((_, reject) => {
            setTimeout(() => reject(new Error('WaitToken Timeout')), timeoutMs);
        });
        return Promise.race([this._promise, timeoutPromise]);
    }

    // 使其支持 await token
    then<TResult1 = any, TResult2 = never>(
        onfulfilled?: ((value: any) => TResult1 | PromiseLike<TResult1>) | undefined | null,
        onrejected?: ((reason: any) => TResult2 | PromiseLike<TResult2>) | undefined | null
    ): Promise<TResult1 | TResult2> {
        return this._promise.then(onfulfilled, onrejected);
    }
}

File: src/lua/radio/wait_token.lua
-- WaitToken implementation in Lua

local WaitToken = {}
WaitToken.__index = WaitToken

function WaitToken.new(wait_cmd, time_base_id)
    local self = setmetatable({}, WaitToken)
    self.wait_cmd = wait_cmd
    self.time_base_id = time_base_id
    self.response = nil
    self.processed_response = nil
    self._done = false
    self._post_processor = nil
    return self
end

function WaitToken:is_done()
    return self._done
end

function WaitToken:set_post_processor(processor)
    self._post_processor = processor
    return self
end

function WaitToken:_apply_post_processor(data)
    local result
    if self._post_processor then
        result = self._post_processor(data)
    else
        result = data
    end
    self.processed_response = result
    return result
end

function WaitToken:complete(data)
    self.response = data
    self:_apply_post_processor(data)
    self._done = true
    -- In Lua, we don't have a standard Event object.
    -- If using a specific framework, you'd trigger an event here.
end

function WaitToken:wait(timeout, client)
    timeout = timeout or 3.0
    local start_time = os.clock()
    while not self._done do
        if os.clock() - start_time > timeout then
            return nil
        end
        if client and client.step then
            client:step()
        end
    end
    return self.processed_response
end

return {
    WaitToken = WaitToken
}
